File: python/sketchatone/cli/strum_event_viewer.py
#!/usr/bin/env python3
"""
Strum Event Viewer CLI

A CLI tool that reads tablet events directly via HID and displays strum events.
Extends blankslate's TabletReaderBase for direct device access.

Usage:
    python -m sketchatone.cli.strum_event_viewer --config path/to/config.json
    python -m sketchatone.cli.strum_event_viewer --config path/to/config.json --strummer-config path/to/strummer.json
    python -m sketchatone.cli.strum_event_viewer --config path/to/config.json --mock
"""

import argparse
import sys
import os
import time
from typing import Optional, Dict, Any, List

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from sketchatone.strummer.strummer import Strummer
from sketchatone.models.strummer_config import StrummerConfig
from sketchatone.models.note import Note, NoteObject

# Import blankslate's TabletReaderBase
try:
    from blankslate.cli.tablet_reader_base import TabletReaderBase, Colors, colored
except ImportError:
    print("Error: blankslate package not found.")
    print("Make sure blankslate is installed: pip install -e ../blankslate/python")
    sys.exit(1)


def print_strummer_info(strummer_config: StrummerConfig):
    """Print strummer-specific configuration info"""
    print(colored('Pressure Threshold: ', Colors.CYAN) +
          colored(str(strummer_config.pressure_threshold), Colors.WHITE))
    print(colored('Notes: ', Colors.CYAN) +
          colored(', '.join(strummer_config.notes), Colors.WHITE))
    if strummer_config.chord:
        print(colored('Chord: ', Colors.CYAN) +
              colored(strummer_config.chord, Colors.WHITE))
    print()


def create_bar(value: float, max_val: float, width: int) -> str:
    """Create a progress bar"""
    filled = int((value / max_val) * width) if max_val > 0 else 0
    filled = min(filled, width)
    empty = width - filled
    return colored('█' * filled, Colors.GREEN) + colored('░' * empty, Colors.GRAY)


def format_note(note: NoteObject) -> str:
    """Format a note for display"""
    secondary_marker = colored('*', Colors.GRAY) if note.secondary else ""
    return colored(f"{note.notation}{note.octave}", Colors.WHITE) + secondary_marker


def print_strum_event(event: Dict[str, Any], strummer: Strummer):
    """Print a strum event in a formatted way"""
    event_type = event.get('type', 'unknown')

    if event_type == 'strum':
        notes = event.get('notes', [])
        note_strs = []
        for note_data in notes:
            note = note_data.get('note')
            velocity = note_data.get('velocity', 0)
            if note:
                note_str = format_note(note)
                vel_bar = create_bar(velocity, 127, 10)
                note_strs.append(f"{note_str} vel:{velocity:3d} {vel_bar}")

        print(colored('♪ STRUM', Colors.GREEN, bold=True))
        for note_str in note_strs:
            print(f"  {note_str}")
        print()

    elif event_type == 'release':
        velocity = event.get('velocity', 0)
        print(colored('↑ RELEASE', Colors.YELLOW) + f" (last velocity: {velocity})")
        print()


def strip_ansi(text: str) -> str:
    """Strip ANSI escape codes from text to get visible length"""
    import re
    return re.sub(r'\033\[[0-9;]*m', '', text)


def pad_line(content: str, target_len: int) -> str:
    """Pad line content accounting for ANSI codes"""
    visible_len = len(strip_ansi(content))
    padding = max(0, target_len - visible_len)
    return content + ' ' * padding


def print_live_dashboard(
    strummer: Strummer,
    x: float,
    y: float,
    pressure: float,
    state: str,
    last_event: Optional[Dict[str, Any]],
    packet_count: int
):
    """Print a live dashboard view with visual string representation"""
    # ANSI codes
    HIDE_CURSOR = '\033[?25l'
    MOVE_HOME = '\033[H'
    CLEAR_LINE = '\033[2K'

    lines = []
    border = colored('│', Colors.CYAN, bold=True)
    box_width = 63

    # Header
    lines.append(colored('┌' + '─' * box_width + '┐', Colors.CYAN, bold=True))
    lines.append(border + colored('                      STRUM EVENT VIEWER                        ', Colors.WHITE, bold=True) + border)
    lines.append(colored('├' + '─' * box_width + '┤', Colors.CYAN, bold=True))

    # Packet counter and state
    packet_str = f"Packets: {packet_count}"
    state_color = Colors.GREEN if state == 'contact' else Colors.YELLOW if state == 'hover' else Colors.GRAY
    state_str = f"State: {colored(state, state_color, bold=True)}"
    lines.append(border + ' ' + pad_line(f"{packet_str}     {state_str}", box_width - 1) + border)

    # Pressure bar
    p_pct = f"{pressure * 100:.0f}%".rjust(4)
    pressure_bar = create_bar(pressure, 1.0, 20)
    threshold = strummer.pressure_threshold
    is_pressed = pressure >= threshold
    pressure_label = colored('PRESSED', Colors.GREEN, bold=True) if is_pressed else '       '
    lines.append(border + ' ' + pad_line(f"Pressure: {p_pct} {pressure_bar} {pressure_label}", box_width - 1) + border)

    lines.append(colored('├' + '─' * box_width + '┤', Colors.CYAN, bold=True))

    # String visualization
    num_strings = len(strummer.notes)
    if num_strings > 0:
        string_width = 1.0 / num_strings
        current_string = min(int(x / string_width), num_strings - 1) if string_width > 0 else 0

        # Calculate spacing for strings
        string_spacing = (box_width - 4) // num_strings
        total_width = string_spacing * num_strings
        left_pad = (box_width - total_width) // 2

        # String characters
        STRING_IDLE = '│'
        STRING_HOVER = '┃'
        STRING_STRUM = '╋'

        # Build string rows (5 rows for visual height)
        for row in range(5):
            row_content = ' ' * left_pad
            for i in range(num_strings):
                is_current_string = i == current_string
                is_strummed = is_current_string and is_pressed
                is_hovered = is_current_string and not is_pressed

                if is_strummed:
                    # Strummed - show vibrating string
                    string_char = colored(STRING_STRUM if row == 2 else '║', Colors.GREEN, bold=True)
                elif is_hovered:
                    # Hovered - highlight
                    string_char = colored(STRING_HOVER, Colors.YELLOW)
                else:
                    # Idle
                    string_char = colored(STRING_IDLE, Colors.GRAY)

                # Center the string character in its column
                col_pad = (string_spacing - 1) // 2
                row_content += ' ' * col_pad + string_char + ' ' * (string_spacing - col_pad - 1)

            lines.append(border + pad_line(row_content, box_width - 1) + border)

        # Note labels row
        note_row = ' ' * left_pad
        for i in range(num_strings):
            note = strummer.notes[i]
            note_label = f"{note.notation}{note.octave}"
            is_current_string = i == current_string
            is_strummed = is_current_string and is_pressed

            if is_strummed:
                colored_label = colored(note_label, Colors.GREEN, bold=True)
            elif is_current_string:
                colored_label = colored(note_label, Colors.YELLOW)
            elif note.secondary:
                colored_label = colored(note_label, Colors.GRAY)
            else:
                colored_label = colored(note_label, Colors.WHITE)

            # Center the note label in its column
            label_len = len(note_label)
            col_pad = (string_spacing - label_len) // 2
            note_row += ' ' * col_pad + colored_label + ' ' * (string_spacing - col_pad - label_len)

        lines.append(border + pad_line(note_row, box_width - 1) + border)

        # Velocity row (show velocity for strummed string)
        vel_row = ' ' * left_pad
        for i in range(num_strings):
            is_current_string = i == current_string
            is_strummed = is_current_string and is_pressed

            if is_strummed:
                # Calculate velocity based on pressure
                velocity = min(127, int(pressure * strummer.velocity_scale * 127))
                vel_label = colored(f"v{velocity:3d}", Colors.CYAN)
            else:
                vel_label = '    '

            col_pad = (string_spacing - 4) // 2
            vel_row += ' ' * col_pad + vel_label + ' ' * (string_spacing - col_pad - 4)

        lines.append(border + pad_line(vel_row, box_width - 1) + border)
    else:
        lines.append(border + pad_line('  No strings configured', box_width - 1) + border)

    # Last event section
    lines.append(colored('├' + '─' * box_width + '┤', Colors.CYAN, bold=True))

    if last_event:
        event_type = last_event.get('type', 'none')
        if event_type == 'strum':
            notes = last_event.get('notes', [])
            note_names = [f"{n['note'].notation}{n['note'].octave}" for n in notes if 'note' in n]
            velocities = [n.get('velocity', 0) for n in notes]
            avg_vel = round(sum(velocities) / len(velocities)) if velocities else 0
            event_line = f"{colored('♪ STRUM', Colors.GREEN, bold=True)} {' '.join(note_names)} {colored(f'vel:{avg_vel}', Colors.CYAN)}"
            lines.append(border + ' ' + pad_line(event_line, box_width - 1) + border)
        elif event_type == 'release':
            vel = last_event.get('velocity', 0)
            event_line = f"{colored('↑ RELEASE', Colors.YELLOW)} {colored(f'(velocity: {vel})', Colors.GRAY)}"
            lines.append(border + ' ' + pad_line(event_line, box_width - 1) + border)
        else:
            lines.append(border + ' ' + pad_line(colored('Waiting for strum...', Colors.GRAY), box_width - 1) + border)
    else:
        lines.append(border + ' ' + pad_line(colored('Waiting for strum...', Colors.GRAY), box_width - 1) + border)

    lines.append(colored('└' + '─' * box_width + '┘', Colors.CYAN, bold=True))
    lines.append(colored('Press Ctrl+C to stop', Colors.GRAY))

    # Build output
    content = '\n'.join(f"{CLEAR_LINE}{line}" for line in lines) + '\n'
    sys.stdout.write(HIDE_CURSOR + MOVE_HOME + content)
    sys.stdout.flush()


class StrumEventViewer(TabletReaderBase):
    """Strum event viewer that reads directly from tablet via HID"""

    def __init__(
        self,
        config_path: str,
        mock: bool = False,
        strummer_config_path: Optional[str] = None,
        live_mode: bool = False
    ):
        super().__init__(config_path, mock=mock, exit_on_stop=True)
        self.live_mode = live_mode
        self.last_event: Optional[Dict[str, Any]] = None
        self.last_live_update = 0

        # Load or create strummer config
        if strummer_config_path:
            self.strummer_config = StrummerConfig.from_json_file(strummer_config_path)
        else:
            self.strummer_config = StrummerConfig()

        # Create strummer
        self.strummer = Strummer()
        self.strummer.configure(
            pluck_velocity_scale=self.strummer_config.velocity_scale,
            pressure_threshold=self.strummer_config.pressure_threshold
        )

        # Set up notes
        self._setup_notes()

    def _setup_notes(self):
        """Set up the strummer notes from config"""
        notes: List[NoteObject] = []

        if self.strummer_config.chord:
            # Parse chord notation
            chord_notes = Note.parse_chord(self.strummer_config.chord)
            notes = Note.fill_note_spread(
                chord_notes,
                self.strummer_config.lower_spread,
                self.strummer_config.upper_spread
            )
        else:
            # Use explicit notes from config
            for note_str in self.strummer_config.notes:
                notes.append(Note.parse_notation(note_str))

        self.strummer.notes = notes

    def start(self):
        """Start viewing strum events"""
        self.print_header('Strum Event Viewer')
        print_strummer_info(self.strummer_config)

        # Initialize reader
        print(colored('Initializing...', Colors.GRAY))
        self.initialize_reader_sync()

        if not self.reader:
            raise RuntimeError('Reader not initialized')

        # Start reading
        print(colored('Setting up data callback...', Colors.GRAY))
        if hasattr(self.reader, 'start_reading'):
            self.reader.start_reading(lambda data: self.handle_packet(data))

        print(colored('✓ Started reading data', Colors.GREEN))
        print(colored('Press Ctrl+C to stop\n', Colors.GRAY))

        self.is_running = True

        if self.is_mock_mode:
            self.start_mock_gesture_cycle_sync()

        # Keep process alive
        try:
            while self.is_running:
                time.sleep(0.1)
        except KeyboardInterrupt:
            pass
        finally:
            self.stop_sync()

    def handle_packet(self, data: bytes):
        """Handle incoming HID packet"""
        try:
            self.packet_count += 1

            # Process the data using the config
            events = self.process_packet(data)

            # Extract normalized values
            x = float(events.get('x', 0))
            y = float(events.get('y', 0))
            pressure = float(events.get('pressure', 0))
            state = str(events.get('state', 'unknown'))

            # Update strummer bounds (use normalized 0-1 range)
            self.strummer.update_bounds(1.0, 1.0)

            # Process strum
            event = self.strummer.strum(x, pressure)

            if event:
                self.last_event = event
                if not self.live_mode:
                    print_strum_event(event, self.strummer)

            if self.live_mode:
                # Throttle live updates to ~10fps
                now = time.time()
                if now - self.last_live_update >= 0.1 or event:
                    self.last_live_update = now
                    print_live_dashboard(
                        self.strummer,
                        x, y, pressure, state,
                        self.last_event,
                        self.packet_count
                    )
        except Exception as e:
            import traceback
            sys.stderr.write(f"\n[ERROR] Failed to process packet: {e}\n")
            traceback.print_exc(file=sys.stderr)
            sys.stderr.flush()


def main():
    parser = argparse.ArgumentParser(
        description='View strum events from tablet input',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
    # Basic usage with tablet config
    python -m sketchatone.cli.strum_event_viewer -c tablet-config.json

    # With custom strummer config
    python -m sketchatone.cli.strum_event_viewer -c tablet-config.json -s strummer-config.json

    # Live dashboard mode
    python -m sketchatone.cli.strum_event_viewer -c tablet-config.json --live

    # Use mock data for testing
    python -m sketchatone.cli.strum_event_viewer -c tablet-config.json --mock
"""
    )

    parser.add_argument(
        '-c', '--config',
        required=True,
        help='Path to tablet config JSON file'
    )

    parser.add_argument(
        '-s', '--strummer-config',
        help='Path to strummer config JSON file'
    )

    parser.add_argument(
        '-l', '--live',
        action='store_true',
        help='Live dashboard mode (updates in place)'
    )

    parser.add_argument(
        '-m', '--mock',
        action='store_true',
        help='Use mock data instead of real device'
    )

    args = parser.parse_args()

    # Validate config file exists
    if not os.path.exists(args.config):
        print(colored(f'Error: Config file not found: {args.config}', Colors.RED))
        sys.exit(1)

    if args.strummer_config and not os.path.exists(args.strummer_config):
        print(colored(f'Error: Strummer config file not found: {args.strummer_config}', Colors.RED))
        sys.exit(1)

    viewer = None
    try:
        viewer = StrumEventViewer(
            config_path=args.config,
            mock=args.mock,
            strummer_config_path=args.strummer_config,
            live_mode=args.live
        )

        # Run synchronously - no asyncio needed
        viewer.start()
    except KeyboardInterrupt:
        print(colored('\n\nShutdown signal received...', Colors.YELLOW))
        if viewer:
            viewer.stop_sync()
        # Show cursor again
        sys.stdout.write('\033[?25h')
        sys.stdout.flush()
        print(colored('\n✓ Exited cleanly', Colors.GREEN))
        sys.exit(0)
    except Exception as error:
        print(colored('Error: ', Colors.RED) + str(error))
        sys.exit(1)


if __name__ == '__main__':
    main()

File: src/cli/strum-event-viewer.ts
#!/usr/bin/env node
/**
 * Strum Event Viewer CLI
 *
 * A CLI tool that reads tablet events directly via HID and displays strum events.
 * Extends blankslate's TabletReaderBase for direct device access.
 *
 * Usage:
 *   npm run strum-events
 *   npm run strum-events -- --config ./configs/
 *   npm run strum-events -- --config path/to/config.json
 *   npm run strum-events -- --config path/to/config.json --strummer-config path/to/strummer.json
 *   npm run strum-events -- --config path/to/config.json --mock
 */

import chalk from 'chalk';
import { Command } from 'commander';
import * as fs from 'fs';
import * as path from 'path';

// Import from blankslate CLI modules
import { TabletReaderBase, type TabletReaderOptions, normalizeTabletEvent, resolveConfigPath } from 'blankslate/cli/tablet-reader-base.js';

// Default config directory
const DEFAULT_CONFIG_DIR = './public/configs';
import { Strummer, type StrummerEvent, type StrumNoteData } from '../core/strummer.js';
import { StrummerConfig } from '../models/strummer-config.js';
import { Note, type NoteObject } from '../models/note.js';

/**
 * Print strummer-specific configuration info
 */
function printStrummerInfo(strummerConfig: StrummerConfig): void {
  console.log(chalk.cyan('Pressure Threshold: ') + chalk.white(strummerConfig.pressureThreshold.toString()));
  console.log(chalk.cyan('Notes: ') + chalk.white(strummerConfig.notes.join(', ')));
  if (strummerConfig.chord) {
    console.log(chalk.cyan('Chord: ') + chalk.white(strummerConfig.chord));
  }
  console.log();
}

/**
 * Create a progress bar
 */
function createBar(value: number, maxVal: number, width: number): string {
  const clampedValue = Math.max(0, value);
  const filled = maxVal > 0 ? Math.min(Math.floor((clampedValue / maxVal) * width), width) : 0;
  const empty = width - filled;
  return chalk.green('█'.repeat(filled)) + chalk.gray('░'.repeat(empty));
}

/**
 * Format a note for display
 */
function formatNote(note: NoteObject): string {
  const secondaryMarker = note.secondary ? chalk.gray('*') : '';
  return chalk.white(`${note.notation}${note.octave}`) + secondaryMarker;
}

/**
 * Print a strum event in a formatted way
 */
function printStrumEvent(event: StrummerEvent): void {
  if (event.type === 'strum') {
    const noteStrs: string[] = [];
    for (const noteData of event.notes) {
      const noteStr = formatNote(noteData.note);
      const velBar = createBar(noteData.velocity, 127, 10);
      noteStrs.push(`${noteStr} vel:${noteData.velocity.toString().padStart(3)} ${velBar}`);
    }

    console.log(chalk.green.bold('♪ STRUM'));
    for (const noteStr of noteStrs) {
      console.log(`  ${noteStr}`);
    }
    console.log();
  } else if (event.type === 'release') {
    console.log(chalk.yellow('↑ RELEASE') + ` (last velocity: ${event.velocity})`);
    console.log();
  }
}

/**
 * Print a live dashboard view with visual string representation
 */
function printLiveDashboard(
  strummer: Strummer,
  x: number,
  y: number,
  pressure: number,
  state: string,
  lastEvent: StrummerEvent | null,
  packetCount: number
): void {
  // ANSI codes
  const HIDE_CURSOR = '\x1b[?25l';
  const MOVE_HOME = '\x1b[H';
  const CLEAR_LINE = '\x1b[2K';

  const lines: string[] = [];
  const border = chalk.cyan.bold('│');
  const boxWidth = 63;

  // Helper to pad line content (accounting for ANSI codes)
  const padLine = (content: string, targetLen: number): string => {
    // Strip ANSI codes to get actual visible length
    const visibleLen = content.replace(/\x1b\[[0-9;]*m/g, '').length;
    const padding = Math.max(0, targetLen - visibleLen);
    return content + ' '.repeat(padding);
  };

  // Header
  lines.push(chalk.cyan.bold('┌' + '─'.repeat(boxWidth) + '┐'));
  lines.push(border + chalk.white.bold('                      STRUM EVENT VIEWER                        ') + border);
  lines.push(chalk.cyan.bold('├' + '─'.repeat(boxWidth) + '┤'));

  // Packet counter and state
  const packetStr = `Packets: ${packetCount}`;
  const stateColor = state === 'contact' ? chalk.green.bold : state === 'hover' ? chalk.yellow : chalk.gray;
  const stateStr = `State: ${stateColor(state)}`;
  lines.push(border + ' ' + padLine(`${packetStr}     ${stateStr}`, boxWidth - 1) + border);

  // Pressure bar
  const pPct = `${(pressure * 100).toFixed(0)}%`.padStart(4);
  const pressureBar = createBar(pressure, 1.0, 20);
  const isPressed = pressure >= strummer.pressureThreshold;
  const pressureLabel = isPressed ? chalk.green.bold('PRESSED') : chalk.gray('       ');
  lines.push(border + ' ' + padLine(`Pressure: ${pPct} ${pressureBar} ${pressureLabel}`, boxWidth - 1) + border);

  lines.push(chalk.cyan.bold('├' + '─'.repeat(boxWidth) + '┤'));

  // String visualization
  const numStrings = strummer.notes.length;
  if (numStrings > 0) {
    const stringWidth = 1.0 / numStrings;
    const currentString = stringWidth > 0 ? Math.min(Math.floor(x / stringWidth), numStrings - 1) : 0;

    // Calculate spacing for strings
    const stringSpacing = Math.floor((boxWidth - 4) / numStrings);
    const totalWidth = stringSpacing * numStrings;
    const leftPad = Math.floor((boxWidth - totalWidth) / 2);

    // String characters
    const STRING_IDLE = '│';
    const STRING_HOVER = '┃';
    const STRING_STRUM = '╋';

    // Build string rows (5 rows for visual height)
    for (let row = 0; row < 5; row++) {
      let rowContent = ' '.repeat(leftPad);
      for (let i = 0; i < numStrings; i++) {
        const isCurrentString = i === currentString;
        const isStrummed = isCurrentString && isPressed;
        const isHovered = isCurrentString && !isPressed;

        let stringChar: string;
        if (isStrummed) {
          // Strummed - show vibrating string
          stringChar = row === 2 ? chalk.green.bold(STRING_STRUM) : chalk.green.bold('║');
        } else if (isHovered) {
          // Hovered - highlight
          stringChar = chalk.yellow(STRING_HOVER);
        } else {
          // Idle
          stringChar = chalk.gray(STRING_IDLE);
        }

        // Center the string character in its column
        const colPad = Math.floor((stringSpacing - 1) / 2);
        rowContent += ' '.repeat(colPad) + stringChar + ' '.repeat(stringSpacing - colPad - 1);
      }
      lines.push(border + padLine(rowContent, boxWidth - 1) + border);
    }

    // Note labels row
    let noteRow = ' '.repeat(leftPad);
    for (let i = 0; i < numStrings; i++) {
      const note = strummer.notes[i];
      const noteLabel = `${note.notation}${note.octave}`;
      const isCurrentString = i === currentString;
      const isStrummed = isCurrentString && isPressed;

      let coloredLabel: string;
      if (isStrummed) {
        coloredLabel = chalk.green.bold(noteLabel);
      } else if (isCurrentString) {
        coloredLabel = chalk.yellow(noteLabel);
      } else {
        coloredLabel = note.secondary ? chalk.gray.dim(noteLabel) : chalk.white(noteLabel);
      }

      // Center the note label in its column
      const labelLen = noteLabel.length;
      const colPad = Math.floor((stringSpacing - labelLen) / 2);
      noteRow += ' '.repeat(colPad) + coloredLabel + ' '.repeat(stringSpacing - colPad - labelLen);
    }
    lines.push(border + padLine(noteRow, boxWidth - 1) + border);

    // Velocity row (show velocity for strummed string)
    let velRow = ' '.repeat(leftPad);
    for (let i = 0; i < numStrings; i++) {
      const isCurrentString = i === currentString;
      const isStrummed = isCurrentString && isPressed;

      let velLabel: string;
      if (isStrummed) {
        // Calculate velocity based on pressure
        const velocity = Math.min(127, Math.floor(pressure * strummer.velocityScale * 127));
        velLabel = chalk.cyan(`v${velocity.toString().padStart(3)}`);
      } else {
        velLabel = '    ';
      }

      const colPad = Math.floor((stringSpacing - 4) / 2);
      velRow += ' '.repeat(colPad) + velLabel + ' '.repeat(stringSpacing - colPad - 4);
    }
    lines.push(border + padLine(velRow, boxWidth - 1) + border);
  } else {
    lines.push(border + padLine('  No strings configured', boxWidth - 1) + border);
  }

  // Last event section
  lines.push(chalk.cyan.bold('├' + '─'.repeat(boxWidth) + '┤'));

  if (lastEvent) {
    if (lastEvent.type === 'strum') {
      const noteNames = lastEvent.notes.map((n: StrumNoteData) => `${n.note.notation}${n.note.octave}`);
      const avgVel = Math.round(lastEvent.notes.reduce((sum, n) => sum + n.velocity, 0) / lastEvent.notes.length);
      const eventLine = `${chalk.green.bold('♪ STRUM')} ${noteNames.join(' ')} ${chalk.cyan(`vel:${avgVel}`)}`;
      lines.push(border + ' ' + padLine(eventLine, boxWidth - 1) + border);
    } else if (lastEvent.type === 'release') {
      const eventLine = `${chalk.yellow('↑ RELEASE')} ${chalk.gray(`(velocity: ${lastEvent.velocity})`)}`;
      lines.push(border + ' ' + padLine(eventLine, boxWidth - 1) + border);
    }
  } else {
    lines.push(border + ' ' + padLine(chalk.gray('Waiting for strum...'), boxWidth - 1) + border);
  }

  lines.push(chalk.cyan.bold('└' + '─'.repeat(boxWidth) + '┘'));
  lines.push(chalk.gray('Press Ctrl+C to stop'));

  // Build output
  const content = lines.map((line) => `${CLEAR_LINE}${line}`).join('\n') + '\n';
  process.stdout.write(HIDE_CURSOR + MOVE_HOME + content);
}

/**
 * Strum event viewer that reads directly from tablet via HID
 */
class StrumEventViewer extends TabletReaderBase {
  private liveMode: boolean;
  private lastEvent: StrummerEvent | null = null;
  private lastLiveUpdate = 0;
  private strummerConfig: StrummerConfig;
  private keepAliveInterval: ReturnType<typeof setInterval> | null = null;
  strummer: Strummer;

  constructor(
    configPath: string,
    options: TabletReaderOptions & {
      strummerConfigPath?: string;
      liveMode?: boolean;
    } = {}
  ) {
    super(configPath, options);
    this.liveMode = options.liveMode ?? false;

    // Load or create strummer config
    if (options.strummerConfigPath) {
      this.strummerConfig = StrummerConfig.fromJsonFile(options.strummerConfigPath);
    } else {
      this.strummerConfig = new StrummerConfig();
    }

    // Create strummer
    this.strummer = new Strummer();
    this.strummer.configure(this.strummerConfig.velocityScale, this.strummerConfig.pressureThreshold);

    // Set up notes
    this.setupNotes();
  }

  private setupNotes(): void {
    let notes: NoteObject[] = [];

    if (this.strummerConfig.chord) {
      // Parse chord notation
      const chordNotes = Note.parseChord(this.strummerConfig.chord);
      notes = Note.fillNoteSpread(chordNotes, this.strummerConfig.lowerSpread, this.strummerConfig.upperSpread);
    } else {
      // Use explicit notes from config
      for (const noteStr of this.strummerConfig.notes) {
        notes.push(Note.parseNotation(noteStr));
      }
    }

    this.strummer.notes = notes;
  }

  protected handlePacket(data: Uint8Array): void {
    try {
      this.packetCount++;

      // Process the data using the config
      const events = this.processPacket(data);

      // Extract normalized values
      const normalized = normalizeTabletEvent(events);
      const { x, y, pressure, state } = normalized;

      // Update strummer bounds (use normalized 0-1 range)
      this.strummer.updateBounds(1.0, 1.0);

      // Process strum
      const event = this.strummer.strum(x, pressure);

      if (event) {
        this.lastEvent = event;
        if (!this.liveMode) {
          printStrumEvent(event);
        }
      }

      if (this.liveMode) {
        // Throttle live updates to ~10fps
        const now = Date.now() / 1000;
        if (now - this.lastLiveUpdate >= 0.1 || event) {
          this.lastLiveUpdate = now;
          printLiveDashboard(this.strummer, x, y, pressure, state, this.lastEvent, this.packetCount);
        }
      }
    } catch (e) {
      const error = e as Error;
      process.stderr.write(`\n[ERROR] Failed to process packet: ${error.message}\n`);
      if (error.stack) {
        process.stderr.write(error.stack + '\n');
      }
    }
  }

  async start(): Promise<void> {
    this.printHeader('Strum Event Viewer');
    printStrummerInfo(this.strummerConfig);

    // Initialize reader
    console.log(chalk.gray('Initializing...'));
    await this.initializeReader();

    if (!this.reader) {
      throw new Error('Reader not initialized');
    }

    // Start reading
    console.log(chalk.gray('Setting up data callback...'));
    this.reader.startReading((data) => {
      this.handlePacket(data);
    });

    console.log(chalk.green('✓ Started reading data'));
    console.log(chalk.gray('Press Ctrl+C to stop\n'));

    if (this.isMockMode) {
      this.startMockGestureCycle();
    }

    // Keep the process alive (HID reader alone doesn't keep Node.js event loop active)
    this.keepAliveInterval = setInterval(() => {}, 1000);

    // Set up shutdown handlers
    this.setupShutdownHandlers();
  }

  async stop(): Promise<void> {
    // Clear keep-alive interval
    if (this.keepAliveInterval) {
      clearInterval(this.keepAliveInterval);
      this.keepAliveInterval = null;
    }

    await super.stop();
  }
}

async function main(): Promise<void> {
  const program = new Command();

  program
    .name('strum-events')
    .description('View strum events from tablet input')
    .option('-c, --config <path>', 'Path to tablet config JSON file or directory (auto-detects from ./public/configs if not provided)')
    .option('-s, --strummer-config <path>', 'Path to strummer config JSON file')
    .option('-l, --live', 'Live dashboard mode (updates in place)')
    .option('-m, --mock', 'Use mock data instead of real device')
    .addHelpText(
      'after',
      `
Examples:
  # Auto-detect tablet from default config directory
  npm run strum-events

  # Auto-detect tablet from specific directory
  npm run strum-events -- -c ./configs/

  # Basic usage with tablet config
  npm run strum-events -- -c tablet-config.json

  # With custom strummer config
  npm run strum-events -- -c tablet-config.json -s strummer-config.json

  # Live dashboard mode
  npm run strum-events -- -c tablet-config.json --live

  # Use mock data for testing
  npm run strum-events -- -c tablet-config.json --mock
`
    );

  program.parse();

  const options = program.opts<{
    config?: string;
    strummerConfig?: string;
    live?: boolean;
    mock?: boolean;
  }>();

  // Resolve tablet config path (handles auto-detection from directory)
  const configPath = resolveConfigPath(options.config ?? DEFAULT_CONFIG_DIR, DEFAULT_CONFIG_DIR);

  if (options.strummerConfig) {
    const strummerConfigPath = path.resolve(options.strummerConfig);
    if (!fs.existsSync(strummerConfigPath)) {
      console.error(chalk.red(`Error: Strummer config file not found: ${strummerConfigPath}`));
      process.exit(1);
    }
  }

  let viewer: StrumEventViewer | null = null;
  try {
    viewer = new StrumEventViewer(configPath, {
      mock: options.mock,
      strummerConfigPath: options.strummerConfig ? path.resolve(options.strummerConfig) : undefined,
      liveMode: options.live,
    });

    await viewer.start();
  } catch (e) {
    const error = e as Error;
    console.error(chalk.red('Error: ') + error.message);
    // Show cursor again
    process.stdout.write('\x1b[?25h');
    process.exit(1);
  }
}

// Only run main() when this file is executed directly (not imported)
import { fileURLToPath } from 'url';
import * as nodePath from 'path';
const __filename = fileURLToPath(import.meta.url);
const isMainModule = nodePath.resolve(process.argv[1]) === __filename;
if (isMainModule) {
  main().catch((error) => {
    console.error(chalk.red('Fatal error:'), error);
    process.exit(1);
  });
}

export { StrumEventViewer, printStrummerInfo, createBar, formatNote, printStrumEvent, printLiveDashboard };